cpp/include/cuml/neighbors/knn.hpp

```diff
@@ -36,36 +38,87 @@ enum MetricType {
   METRIC_Correlation
 };
 
+struct knnIndex {
+  faiss::gpu::StandardGpuResources *gpu_res;
```

Comment:
Either before or shortly after this PR is merged, we need to update FAISS in cuML and use their new pluggable memory manager feature (https://github.com/facebookresearch/faiss/issues/1203). While the brute-force computation uses only a very small workspace, the approximate `Index` variants put FAISS in complete control of the memory space of the index (through the `StandardGPUResources`).